src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,270 @@
+        /// Remove service node from swarm
+        auto &swarm_sn_vec = swarm_to_snodes.at(random_sn_swarm_id);
+        swarm_sn_vec.erase(std::remove(swarm_sn_vec.begin(), swarm_sn_vec.end(), random_sn_pk), swarm_sn_vec.end());
+        if (swarm_sn_vec.size() <= EXCESS_BASE)
+        {
+          /// Remove the swarm's snodes from the pool
+          excess_pool.erase(std::remove_if(excess_pool.begin(),
+                                          excess_pool.end(),
+                                          [&](const excess_pool_snode &excess) {
+                                            return excess.swarm_id == random_sn_swarm_id;
+                                          }),
+                            excess_pool.end());
+        }
+        else
+        {
+          /// Remove only the one snode from the pool
+          excess_pool.erase(excess_pool.begin() + idx);
+        }
+
+        /// Add to new swarm
+        new_swarm_snodes.push_back(random_sn_pk);
+      }
+      const auto new_swarm_id = get_new_swarm_id(mt, swarm_to_snodes);
+      swarm_to_snodes.insert({new_swarm_id, std::move(new_swarm_snodes)});
+      LOG_PRINT_L2("Created new swarm from excess: " << new_swarm_id);
+    }
+  }
+
+  prod_static void calc_swarm_sizes(const swarm_snode_map_t &swarm_to_snodes, std::vector<swarm_size> &sorted_swarm_sizes)
+  {
+    sorted_swarm_sizes.clear();
+    for (const auto &entry : swarm_to_snodes)
+    {
+      sorted_swarm_sizes.push_back({entry.first, entry.second.size()});
+    }
+    std::sort(sorted_swarm_sizes.begin(),
+              sorted_swarm_sizes.end(),
+              [](const swarm_size &a, const swarm_size &b) {
+                return a.size < b.size;
+              });
+  }
+
+  prod_static void assign_snodes(const std::vector<crypto::public_key> &snode_pubkeys, swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
+  {
+    std::vector<swarm_size> sorted_swarm_sizes;
+    for (const auto &sn_pk : snode_pubkeys)
+    {
+      calc_swarm_sizes(swarm_to_snodes, sorted_swarm_sizes);
+      const size_t percentile_index = FILL_SWARM_LOWER_PERCENTILE * (sorted_swarm_sizes.size() - 1) / 100;
+      const size_t percentile_value = sorted_swarm_sizes.at(percentile_index).size;
+      /// Find last occurence of percentile_value
+      size_t upper_index = sorted_swarm_sizes.size() - 1;
+      for (size_t i = 0; i < sorted_swarm_sizes.size(); ++i)
+      {
+        if (sorted_swarm_sizes[i].size > percentile_value)
+        {
+          /// Would never happen for i == 0
+          upper_index = i - 1;
+          break;
+        }
+      }
+      const size_t random_idx = uniform_distribution_portable(mt, upper_index + 1);
+      const swarm_id_t swarm_id = sorted_swarm_sizes[random_idx].swarm_id;
+      swarm_to_snodes.at(swarm_id).push_back(sn_pk);
+      /// run the excess/threshold round after each additional snode
+      create_new_swarm_from_excess(swarm_to_snodes, mt);
```

Comment:
I wonder if this needs to be called every time, can't you just leave the one call you already have after "Robin Hood stealing"?

Reply:
By doing so, the algorithm can process any arbitrary number of registrations from scratch. Just thought it would make the algorithm more robust to be independent of the concurrent number of registrations.

Reply:
The problem is that you shouldn't try to create a new swarm before you check no other existing swarm is starving.

Reply:
Good point. I guess I could return early in `create_new_swarm_form_excess` if there is a starving swarm.

Reply:
I still don't think you need to call it this early. I don't think you will ever create a new swarm out of a dissolved swarm (because if you had that many nodes you wouldn't have to dissolve it in the first place), so the fact that `assign_snodes` is called in step 4 shouldn't matter. What problem does this solve compared to just creating new swarms in step 3?

Reply:
So to clarify things up:
`assign_snodes` is called for both:
- new registrations
- snodes that have been pulled out of a decommissioned swarm and need to be re-assigned.

The reason `assign_snodes` calls `create_new_swarm_from_excess` is to improve behaviour for new bulk registration, and make the algorithm work regardless of whether the new snodes were registered all at once or in separate blocks.

The use case is when we want to handle registering a large amount of snodes (imagine we wanted to reset the mapping and register all the current snodes at once, without having to rescan the blockchain). So the existing map is empty and we have 400+ snodes to be registered. Using the algorithm without calling `create_new_swarm_form_excess`, a first swarm would be created (step 0) then filled with *all* the snodes (step 2). Then step 3 would create additional swarms of size IDEAL_SWARM_SIZE, one by one. But by doing so, you don't end up with a good uniformity: when the excess drops below the threshold, the first swarm would still be 2-3 time bigger than the average swarm (I have observed that in the unit tests), and some of the swarms would be a couple of snodes below IDEAL_SWARM_SIZE because they participated in a subsequent new swarm creation. 

So by calling `create_new_swarm_form_excess` after assigning each new registration, you basically start with 1 empty swarm (step 0) then as you add new snodes in step 1, you create more swarms as required. Then step 3 doesn't do anything since all swarms are nice and uniform.

This allows the algorithm to work normally, as if all the registrations had happen in separate blocks...

Reply:
Mark as resolved as there doesn't seem to be an easy alternative to what you are proposing.